airflow/io/xcom.py

```diff
@@ -0,0 +1,111 @@
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import json
+import sys
+import uuid
+from typing import TYPE_CHECKING, Any, TypeVar
+from urllib.parse import urlsplit
+
+from airflow.configuration import conf
+from airflow.io.path import ObjectStoragePath
+from airflow.models.xcom import BaseXCom
+from airflow.utils.json import XComDecoder, XComEncoder
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+    from airflow.models import XCom
+
+T = TypeVar("T")
+
+
+class XComObjectStoreBackend(BaseXCom):
+    """XCom backend that stores data in an object store."""
+
+    path = conf.get("core", "xcom_objectstore_path", fallback="")
+
+    @staticmethod
+    def _get_key(data: str) -> str:
+        path = conf.get("core", "xcom_objectstore_path", fallback="")
+        p = ObjectStoragePath(path)
+
+        url = urlsplit(data)
+        if url.scheme:
+            k = ObjectStoragePath(data)
+
+            if k.is_relative_to(p) is False:
+                raise ValueError(f"Invalid key: {data}")
+            else:
+                return data.replace(path, "", 1).lstrip("/")
+
+        raise ValueError(f"Not a valid url: {data}")
+
+    @staticmethod
+    def serialize_value(
+        value: T,
+        *,
+        key: str | None = None,
+        task_id: str | None = None,
+        dag_id: str | None = None,
+        run_id: str | None = None,
+        map_index: int | None = None,
+    ) -> bytes | str:
+        s_val = json.dumps(value, cls=XComEncoder).encode("utf-8")
+        path = conf.get("core", "xcom_objectstore_path", fallback="")
+        compression = conf.get("core", "xcom_objectstore_compression", fallback=None)
+        threshold = conf.getint("core", "xcom_objectstore_threshold", fallback=-1)
+
+        if path and -1 < threshold < sys.getsizeof(value):
+            p = ObjectStoragePath(path) / f"{run_id}/{task_id}/{str(uuid.uuid4())}"
+
+            if not p.exists():
+                p.parent.mkdir(parents=True, exist_ok=True)
+
+            with p.open("wb", compression=compression) as f:
+                f.write(s_val)
+
+            return BaseXCom.serialize_value(str(p))
+        else:
+            return s_val
+
+    @staticmethod
+    def deserialize_value(
+        result: XCom,
+    ) -> Any:
+        data = BaseXCom.deserialize_value(result)
+        path = conf.get("core", "xcom_objectstore_path", fallback="")
+        try:
+            p = ObjectStoragePath(path) / XComObjectStoreBackend._get_key(data)
```

Comment:
There are two potential problems here:

1) there is quite an extra overhead for small values here - where retrieving of small value always involves extra storage call that could be avoided
2) there is a potential (small probablitly of course) - that there will be a storage object created externally and it might wreak havoc

Maybe another option would be to add a bool field in the Xcom table to distinquish local/remote? Mostly because of 1). I think it would be a way more performant approach for a number of cases. We've already had precedents with secret backends where reaching out to remote before checking things in the DB caused a lot of cost and network overhead. I think we should learn from that.

Reply:
I don't think the overhead is really there. Small values will not result in that extra call, because the key won't be valid and it will raise a ValueError and object storage won't be touched.

Reply:
Yeah. I think you are right, but It's a bit relying on implicit behaviour in this case. I think it should be at least described in a comment here that we know about it and that was designed and not accidental feature

Reply:
Okido will add that.